File: jmix-flowui/flowui-kit/src/main/java/io/jmix/flowui/kit/component/sidepanellayout/JmixSidePanelLayoutCloser.java
/*
 * Copyright 2026 Haulmont.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.jmix.flowui.kit.component.sidepanellayout;

import com.vaadin.flow.component.*;
import com.vaadin.flow.component.dependency.JsModule;
import io.jmix.flowui.kit.icon.JmixFontIcon;
import jakarta.annotation.Nullable;

@Tag("jmix-side-panel-layout-closer")
@JsModule("./src/side-panel-layout/jmix-side-panel-layout-closer.js")
public class JmixSidePanelLayoutCloser extends Component implements HasTheme, Focusable<JmixSidePanelLayoutCloser>,
        HasStyle, HasAriaLabel, HasSize {

    protected Component icon;
    protected Component defaultIcon;
    protected JmixSidePanelLayout sidePanelLayout;

    public JmixSidePanelLayoutCloser() {
        setDefaultIcon(JmixFontIcon.SIDE_PANEL_LAYOUT_CLOSER.create());
        setIcon(null);
    }

    public JmixSidePanelLayout getSidePanelLayout() {
        return sidePanelLayout;
    }

    public void setSidePanelLayout(JmixSidePanelLayout sidePanelLayout) {
        this.sidePanelLayout = sidePanelLayout;

        getElement().executeJs("this.sidePanelElement = $0", sidePanelLayout);
    }

    @Nullable
    public Component getIcon() {
        return icon;
    }

    public void setIcon(@Nullable Component icon) {
        if (icon != null && icon.getElement().isTextNode()) {
            throw new IllegalArgumentException("Text node can't be used as an icon");
        }

        if (this.icon != null) {
            remove(this.icon);
        } else {
            remove(defaultIcon);
        }

        this.icon = icon;

        if (icon == null) {
            add(defaultIcon);
        } else {
            add(icon);
        }

        updateThemeAttribute();
    }

    protected void setDefaultIcon(Component defaultIcon) {
        this.defaultIcon = defaultIcon;
    }

    protected void add(Component... components) {
        for (Component component : components) {
            getElement().appendChild(component.getElement());
        }
    }

    protected void remove(Component... components) {
        for (Component component : components) {
            if (getElement().equals(component.getElement().getParent())) {
                getElement().removeChild(component.getElement());
            }
        }
    }

    protected void updateThemeAttribute() {
        if (getChildren().count() == 1) {
            getThemeNames().add("icon");
        } else {
            getThemeNames().remove("icon");
        }
    }
}

File: jmix-flowui/flowui-kit/src/main/resources/META-INF/frontend/src/side-panel-layout/jmix-side-panel-layout-closer.js
/*
 * Copyright 2026 Haulmont.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { html, LitElement } from 'lit';
import { ButtonMixin } from '@vaadin/button/src/vaadin-button-mixin.js';
import { PolylitMixin } from '@vaadin/component-base/src/polylit-mixin.js';
import { LumoInjectionMixin } from '@vaadin/vaadin-themable-mixin/lumo-injection-mixin.js';
import { defineCustomElement } from '@vaadin/component-base/src/define.js';
import { DirMixin } from '@vaadin/component-base/src/dir-mixin.js';
import { isEmptyTextNode } from '@vaadin/component-base/src/dom-utils.js';
import { ThemableMixin } from '@vaadin/vaadin-themable-mixin/vaadin-themable-mixin.js';
import { jmixSidePanelLayoutCloserStyles } from './styles/jmix-side-panel-layout-closer-base-styles.js';

class JmixSidePanelLayoutCloser extends ButtonMixin(DirMixin(ThemableMixin(PolylitMixin(LumoInjectionMixin(LitElement))))) {

    static get is() {
        return 'jmix-side-panel-layout-closer';
    }

    static get styles() {
        return jmixSidePanelLayoutCloserStyles;
    }

    static get properties() {
        return {
            ariaLabel: {
                type: String,
                reflectToAttribute: true,
            },
            /** @private */
            _showFallbackIcon: {
              type: Boolean,
              value: false,
            },
        };
    }

    render() {
        return html`
            <slot id="slot" @slotchange="${this._toggleFallbackIcon}">
              <div part="icon"></div>
            </slot>
            <div part="icon" ?hidden="${!this._showFallbackIcon}"></div>
        `;
    }

    ready() {
        super.ready();

        this.addEventListener('click', e => this._onClick(e));

        this._toggleFallbackIcon();
    }

    _onClick(e) {
        if (this.sidePanelElement && this.sidePanelElement._closeSidePanel) {
            this.sidePanelElement._closeSidePanel();
        }
    }

    /*
     * @private
     */
    _toggleFallbackIcon() {
      const nodes = this.$.slot.assignedNodes();

      // Show fallback icon if there are 1-2 empty text nodes assigned to the default slot.
      this._showFallbackIcon = nodes.length > 0 && nodes.every((node) => isEmptyTextNode(node));
    }
}

defineCustomElement(JmixSidePanelLayoutCloser);

export { JmixSidePanelLayoutCloser };
